Add a `wait_for_samples` method to the `MovingWindow`

shsms opened the pull request:
Closes #967

shsms committed:
[MovingWindow] Accept a time range in the `count_valid` method

It retains the original behaviour of counting all the valid samples in
the buffer when no time range is specified.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
[MovingWindow] Add a `wait_for_samples` method

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update release notes

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

cwasicki commented on src/frequenz/sdk/timeseries/_ringbuffer/buffer.py:
```diff
@@ -678,13 +678,29 @@ def count_covered(self) -> int:
+        Args:
+            since: The timestamp from which to start counting.  If `None`, the oldest
+                timestamp in the buffer is used.
+            until: The timestamp until which to count.  If `None`, the newest timestamp
```
Better clarify that `until` is inclusive since usually the end timestamp (or index) is exclusive in python.

shsms replied:
done

cwasicki commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -318,6 +320,34 @@ def window(
+            raise ValueError(
+                "The number of samples to wait for must be greater than 0."
+            )
+        if n > self.capacity:
```
Theoretically I don't see why this is required but also don't have a good example where it hurts.

shsms replied:
We are counting the number of items in the buffer, for reporting.  If we have to support n > capacity, we'll need a different implementation.  But I couldn't think of a usecase for allowing n > capacity.

cwasicki commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -318,6 +320,34 @@ def window(
+        start_timestamp = self.newest_timestamp
+        if n < self.capacity:
+            n += self.count_valid(since=start_timestamp)
+        while True:
```
I am not familiar with the Condition concept and cannot follow what is happening here, maybe you could add some explanatory comments here and in the tests below.

shsms replied:
I've added a few lines, hope that's sufficient.

cwasicki commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -318,6 +320,34 @@ def window(
         )
 
+    async def wait_for_samples(self, n: int) -> None:
+        """Wait until the next `n` samples are available in the MovingWindow.
```
I guess this is _valid_ samples? If so, I am actually not sure if we want this or something time-based i.e. allow that not all samples are valid. However, for our current use-case this also works since we would set n=1 anyway.

shsms replied:
That's interesting.  _valid_ means that any data was received.  If a component is missing data, resampler will send `None` and that is not a valid value.

If a component is sending only `None`, should this function return after `n` `None`s are received?  I'm guessing it should?

cwasicki replied:
I see, in many scenarios I wouldn't distinguish between missing or None values. I think it shouldn't return after n Nones but after `n` new time steps which have at least 1 real value. However, we could leave this also for later.

shsms replied:
I've updated it to return after `n` samples are received.  Whether they were valid or not needs to be checked with a call to `count_valid`.  I've also updated the docs to state this.

cwasicki replied:
Oh, I see the confusion, I understood that it triggers when `n` new _output_ samples have been "received", i.e. there are timestamps in the resulting moving window. But this is about _input_ samples. So even if we receive 100 samples, if these are all older than newest timestamp we wouldn't get any new timestamp in the window but updated data points of older timestamps. 

This makes sense to me, would stress that in the doc though, e.g the valid samples part is confusing IMO since this is indeed about the new samples.

cwasicki replied:
Maybe:
```
       """Wait until the next `n` samples have been received in the MovingWindow.

        This function returns after `n` input samples have been received, without considering
        whether the received samples are valid or which timestamp they have. The validity of 
        the samples in the updated moving window can be verified by calling the
        [`count_valid`][frequenz.sdk.timeseries.MovingWindow.count_valid] method.
```

shsms replied:
No, it would wait until there are `n` output samples, but some output samples could be nan.

It does consider the timestamps when the samples are received.  It expects `n` "new" samples to be available in the buffer before it returns.

The current tests only cover cases where there is no resampling in the moving window.  I'll rectify that.

shsms replied:
Updated to clarify that it returns after `n` new samples in the moving window, and not just input samples, as discussed.

shsms committed:
[MovingWindow] Accept a time range in the `count_covered` method

It retains the original behaviour of counting all the valid samples in
the buffer when no time range is specified.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

cwasicki commented on tests/timeseries/test_moving_window.py:
```diff
@@ -210,21 +213,234 @@ async def test_access_empty_window() -> None:
+        await push_logical_meter_data(sender, range(0, 5))
+        await asyncio.sleep(0)
+        # After pushing 5 values, the `wait_for_samples` task should be done.
+        assert task.done()
```
Now looking at the indended usage example here, would it be possible to wrap this into a receiver that sends the moving window content each time the method triggers? We could implement this in downstream apps of course, but I guess in the end this would be the pattern we would use.

shsms replied:
Yup, can do that.

cwasicki replied:
As discussed, don't think we need it for this PR but can be added in future if we prefer that pattern.

llucax commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -318,6 +320,44 @@ def window(
+            ValueError: If `n` is less than or equal to 0 or greater than the capacity
+                of the MovingWindow.
+        """
+        if n <= 0:
+            raise ValueError(
+                "The number of samples to wait for must be greater than 0."
+            )
```
I wonder if returning immediately couldn't be a more useful alternative, in cases where maybe `n` is calculated and could result in 0, in which case you actually don't need to wait for anything. I guess @cwasicki can hint which approach could be more convenient for the use cases he has in mind.

cwasicki replied:
Good idea, that would save a try block

llucax replied:
Not sure about negative values though, as it is not perfectly clear what it means (as opposed to "wait for zero samples" which for me translates more naturally to "don't wait"). Although if we don't raise I think the only possible meaning is to also not wait. If we raise, I guess if you could get a negative you can always use `mw.wait_for_samples(max(0, calculated_n_that_can_be_negative))` so it wouldn't be that bad either.

shsms replied:
done, still raises for < 0.

llucax commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -318,6 +320,44 @@ def window(
+                "The number of samples to wait for must be greater than 0."
+            )
+        if n > self.capacity:
+            raise ValueError(
+                "The number of samples to wait for must be less than or equal to the "
+                + f"capacity of the MovingWindow ({self.capacity})."
+            )
```
This could also just silently wait for `self.capacity` instead, like slicing when you do `[:10]` for an array that has less than 10 items. It can be more confusing if you wanted to wait for 10 and got 5 instead but it is something at least familiar in Python. Again maybe @cwasicki can hint which approach would be more intuitive for data scientists.

cwasicki replied:
I thought of it that if n > capacity the n would still be respected, but calculated in terms of time steps since this was triggered. But I think the different understanding here shows already that this could be confusing and we can limit it until we have a case where we need it.

llucax replied:
Yeah, true, best to start with a safe approach, for my behaviour you can still easily get is via `mw.wait_for_samples(min(n, mw.capacity))` I guess, so it doesn't look like it is adding too much value.

llucax commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -318,6 +320,44 @@ def window(
+            else None
+        )
+        while True:
+            async with self._condition_new_sample:
+                # Every time a new sample is received, this condition gets notified and
+                # will wake up.
+                _ = await self._condition_new_sample.wait()
```
Just thinking out loud, and I don't even think we need to think about it for this PR, but maybe this could be done more efficiently by reversing the logic, and only set the condition when the counter is set (it could be even be a simple `Event` in this case). This would mean we would need to save the "waiters" in the instance, and then notify the appropriate waiter, so maybe that's a bit costly too, but I guess the most common case will be having only one or very few waiters. :thinking:

shsms replied:
The most common case is also with n=1 and resampling_interval=15 minutes.

llucax commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -318,6 +320,44 @@ def window(
             start, end, force_copy=force_copy, fill_value=fill_value
         )
 
+    async def wait_for_samples(self, n: int) -> None:
```
Maybe:
```suggestion
    async def wait_for_samples(self, n: int, /) -> None:
```
as using it as a keyword doesn't really bring anything.

shsms replied:
Not doing it, because what it someone wants to.

llucax commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -355,23 +401,44 @@ async def sink_buffer(sample: Sample[Quantity]) -> None:
-        """
-        Count the number of valid samples in this `MovingWindow`.
+    def count_valid(
+        self, since: datetime | None = None, until: datetime | None = None
```
Here I think is more important to use keyword-only as both have the same time and could be mixed up.
```suggestion
        self, *, since: datetime | None = None, until: datetime | None = None
```

shsms replied:
done

llucax commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -355,23 +401,44 @@ async def sink_buffer(sample: Sample[Quantity]) -> None:
 
-    def count_covered(self) -> int:
+    def count_covered(
+        self, since: datetime | None = None, until: datetime | None = None
```
Here too:
```suggestion
        self, *, since: datetime | None = None, until: datetime | None = None
```

shsms replied:
done

shsms committed:
[MovingWindow] Don't filter out `None` values from resampler

The resampler sends `None` values when there is no source data.  In
this case, the values need to be sent to the buffer, so that it can
update its gaps immediately, instead of waiting until when data is
finally available.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
Also added tests for `wait_for_samples` in moving windows with a resampler.

cwasicki commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -318,6 +320,44 @@ def window(
+    async def wait_for_samples(self, n: int) -> None:
+        """Wait until the next `n` samples are available in the MovingWindow.
+
+        This function returns after `n` samples are available in the MovingWindow,
```
Would say `after n new samples are available`

shsms replied:
done

cwasicki reviewed (dismissed):
LGTM, just minor clarification proposal.

llucax reviewed (approved):
LGTM. I disabled auto-merge to give @cwasicki the opportunity to approve too, as he knows what's needed best.

shsms commented:
He had already approved.
